```{r LIBRARIES, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
library(pacman)
p_load(GenomicRanges, ggplot2, rtracklayer, dplyr, plyranges, ChIPseeker,
TxDb.Mmusculus.UCSC.mm10.knownGene, TxDb.Hsapiens.UCSC.hg38.knownGene,
BSgenome.Mmusculus.UCSC.mm10, BSgenome.Hsapiens.UCSC.hg38, org.Mm.eg.db,
org.Hs.eg.db, annotables, rBLAST, rentrez)
# nolint end
```

```{r PATH_DECLARATION, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
PROJECT     <- paste0("/home/daniele/Desktop/IV_course/II_semester/TF_analysis/")
BIGBEDS     <- paste0(PROJECT, "Input/")
MMUSCULUS   <- paste0(BIGBEDS, "Mus_musculus/")
HSAPIENS    <- paste0(BIGBEDS, "Homo_sapiens/")
INTER_FILES <- paste0(PROJECT, "Intermediate_data/")
SCRIPTS     <- paste0(PROJECT, "Scripts/")
RESULTS     <- paste0(INTER_FILES, "Generated_files")
FASTAS      <- paste0(INTER_FILES, "FASTA/")
FIGURES     <- paste0(PROJECT, "Figures/")

SAMPLES     <- read.csv(file = paste0(INTER_FILES, "sample_key.csv"))
DATABASES   <- paste0(PROJECT, "Databases/")
# nolint end
```

```{r FUNCTION_SOURCING, echo=FALSE, message=FALSE, warning=FALSE}
source(paste0(SCRIPTS, "functions.R"))
source(paste0(SCRIPTS, "plots.R"))
```

```{r BIGBED_READING, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Listing sample BigBed files for each analyzed organism:
bbfiles_mm <- list.files(path = MMUSCULUS, "*bb")
bbfiles_hg <- list.files(path = HSAPIENS, "*bb")

# Making a granges list for each analyzed organism:
grl_mm <- GRangesList()
grl_hg <- GRangesList()

chr_abr <- c(paste0("chr", 1:19), "chrX", "chrY")
len_mm <- length(bbfiles_mm)    # Length of samples for Mus musculus
len_hg <- length(bbfiles_hg)    # Length of samples for Homo sapiens

# Creating GRanges objects and extracting certain chromosomes for Mus musculus:
for (i in 1:len_mm) {
    grl_mm[[i]] <-
        import(paste0(MMUSCULUS, bbfiles_mm[i])) %>%
        filter(seqnames %in% chr_abr)
    names(grl_mm)[i] <- SAMPLES$Graph_names[SAMPLES$Filename == bbfiles_mm[i]]
}

# Creating GRanges objects and extracting certain chromosomes for Homo sapiens.
# These files will be used as control samples to check whether TF binding sites
# were predicted successfully using Homo sapiens annotation:
for (i in 1:len_hg) {
    grl_hg[[i]] <-
        import(paste0(HSAPIENS, bbfiles_hg[i])) %>%
        filter(seqnames %in% chr_abr)
    names(grl_hg)[i] <- SAMPLES$Graph_names[SAMPLES$Filename == bbfiles_hg[i]]
}
# nolint end
```

```{r PWM_READING, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Making a list of analyzed TF PWM matrices from different organisms:
pwm <- get_PWM("GATA3_MOUSE.H11MO.0.A.pwm")
# nolint end
```

```{r MM_PEAK_ANNOTATION, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
mm_annotated_peaks <-
    annotate_peaks(grl_mm, grcm38, TxDb.Mmusculus.UCSC.mm10.knownGene,
                   "org.Mm.eg.db")

write.table(mm_annotated_peaks,
            file = paste0(RESULTS, "/mm_annotated_peaks.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE)
# nolint end
```

```{r}
# nolint start
mmgene_ranges <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)

# Adding 'gene_symbol' column to gene_ranges GRange:
mmgene_ranges$gene_symbol <-
    mapIds(org.Mm.eg.db, keys = mmgene_ranges$gene_id, column = "SYMBOL",
           keytype = "ENTREZID", multiVals = "first")

mm_genes <- GRangesList()

for (sample in seq_along(mm_annotated_peaks)) {
    selected_genes <-
        tolower(unique(mm_annotated_peaks[[sample]]$gene_symbol)) %in%
        tolower(mmgene_ranges$gene_symbol)
                            
    genes <- mmgene_ranges[selected_genes]
    mm_genes[[sample]] <- genes 
}

gene_sequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10, mm_genes[[1]])

# REIKIA PASPARTINTI?
for (sample in 1:length(gene_sequences)) {
    if (is.na(as.character(mm_genes[[1]][sample]$gene_symbol))) {
        names(gene_sequences)[sample] <- "Unknown"
    } else {
        names(gene_sequences)[sample] <-
        as.character(mm_genes[[1]][sample]$gene_symbol)
    }
}

# as.character(mm_genes[[1]][1]$gene_symbol)
# endoapply(names(gene_sequences), function(x) (names(gene_sequences)[x] <-
#             as.character(mm_genes[[1]][x]$gene_symbol)))

# nolint end
```


```{r BLAST_SEARCH, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
human_db <- blast(db = paste0(DATABASES,
                  "Human_genome_DB/GCF_000001405.39_top_level.00"))

gene_sequencess <- gene_sequences
predictions <-
    predict(human_db, gene_sequencess, BLAST_args = "-perc_identity 90",
            custom_format = "qseqid sacc sstart send bitscore length stitle")


# KAI -perc_identity 99, tada jokių atitikimų nesuranda.


sink(paste0(PROJECT, "predictions3.txt"))
print(predictions)
sink()
# write(predictions, paste0(PROJECT, "predictions3.txt"))


# colnames(predictions)[8] <- "Assembly"
# predictions <- predictions %>%
#     group_by(qseqid) %>%
#     filter(row_number() == 1) %>%
#     as.data.frame

predictions <- read.table(paste0(PROJECT, "predictions_3.txt"))
colnames(predictions)

head(predictions)

a <- 
predictions %>%
as.data.frame() %>%
dplyr::select(c(sacc, sstart, send))


 %>%
makeGRangesFromDataFrame()


colnames(a) <- c("seqnames", "start", "end")
a <- makeGRangesFromDataFrame(a)

a <- 
a %>%
  add_column(strand = NA)

for (i in 1:length(rownames(a))) {
    if (a$start[i] > a$end[i]) {
        a$strand <- "-"
        start <- a$end[i]
        end <- a$start[i] 
        a$start[i] <- start
        a$end[i] <- end
    } else {
        a$strand <- "+"
    }
}

head(a)

library(GenomicRanges)

subsetByOverlaps(a, significant_peaks)

ranges(a)
ranges(significant_peaks)

library(dplyr)
library(tibble)

# KODĖL NESURANDA JOKIŲ POZICIJŲ?

# Nuo šios dalies pratęsti!
for (acc in 1:length(predictions$qseqid)) {
    # if (predictions$sstart[acc] >= as.data.frame(ranges(significant_peaks))$start &&
    #     predictions$send[acc] <= as.data.frame(ranges(significant_peaks))$end ||
    #     predictions$sstart[acc] <= as.data.frame(ranges(significant_peaks))$start &&
    #     predictions$send[acc] >= as.data.frame(ranges(significant_peaks))$end) {
    #         print("OOOK")
    #     }
    #     else {print("BAAAAD")}
    # sequence <-
    #     rentrez::entrez_fetch(db = "nucleotide", id = predictions$sacc[acc], 
    #                           rettype = "fasta")
    # write(sequence, paste0(FASTA, predictions$sacc[acc]))
    # IMPROVE ABOVE LINE OF CODE!!! (DO NOT WRITE INTO A FILE!)

    for (i in 1:length(ranges(significant_peaks))) {
        if (predictions$sstart[acc] >= as.data.frame(ranges(significant_peaks)[i])$start &&
        predictions$send[acc] <= as.data.frame(ranges(significant_peaks)[i])$end) {
            print("OOOK")
        }
    }
}

as.data.frame(ranges(significant_peaks)[1])$start

predictions$sacc[1]
length(predictions$qseqid)

sequence_list <- list()

for (seq in 1:length(predictions$sacc)) {
    sequence_list[[seq]] <-
        readDNAStringSet(paste0(PROJECT, predictions$sacc[seq]),
                         format = "fasta")
    names(sequence_list[[seq]]) <- predictions$sacc[seq]
    names(sequence_list)[seq] <- predictions$sacc[seq]
}

# NAUDOTI ŠĮ REZULTATĄ!
pwm_hit_predicted <- count_pwm_hits(sequence_list, pwm, "predicted")

# nolint end
```



```{r SIGNIFICANT_PEAKS}
# nolint start
# q value must be lower than 2 (-log10(0.01) = 2). These peaks are
# significant peaks:
significant_peaks <- grl_hg[[2]][grl_hg[[2]]$q_value < 2, ]

significant_peaks <- list(significant_peaks)
hg_annotated_peaks <-
    annotate_peaks(significant_peaks, grch38, TxDb.Hsapiens.UCSC.hg38.knownGene,
                   "org.Hs.eg.db")

hg_annotated_peaks <- as.data.frame(hg_annotated_peaks)
b <- toupper(unique(hg_annotated_peaks$gene_symbol))

c <- toupper(unique(predictions$qseqid))

write.csv(as.data.frame(b), paste0(PROJECT, "b2.txt"), row.names = FALSE, quote = FALSE)
write.csv(as.data.frame(c), paste0(PROJECT, "c2.txt"), row.names = FALSE, quote = FALSE)
intersect(b, c)
# nolint end
```




















```{r HG_ANNOTATION, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
# Getting positions of Homo sapiens genes:
gene_ranges <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene)

# Adding 'gene_symbol' column to gene_ranges GRange:
gene_ranges$gene_symbol <-
    mapIds(org.Hs.eg.db, keys = gene_ranges$gene_id, column = "SYMBOL",
           keytype = "ENTREZID", multiVals = "first")

write.table(gene_ranges,
            file = paste0(RESULTS, "/hg_annotation_genes.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE)

hg_genes <- GRangesList()

for (sample in seq_along(mm_annotated_peaks)) {
    selected_genes <-
        tolower(unique(mm_annotated_peaks[[sample]]$gene_symbol)) %in%
        tolower(gene_ranges$gene_symbol)
                            
    genes <- gene_ranges[selected_genes]

    # These are the Homo sapiens genes that overlap with annotated Mus musculus
    # peak genes. This information will be used to compare the results with
    # control Homo sapiens samples:
    hg_genes[[sample]] <- genes 
}
# nolint end
```


```{r GENE_COUNTS, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
genes_samples <- data.frame(matrix(ncol = 3, nrow = 0))
colnames(genes_samples) <- c("Sample", "Gene_count", "Percentage")

for (sample in 1:length(mm_annotated_peaks)) {
    genes_in_sample <- length(unique(mm_annotated_peaks[[sample]]$gene_symbol))
    row <-
        c(names(mm_annotated_peaks[sample]), genes_in_sample,
          (genes_in_sample / length(unique(gene_ranges$gene_symbol))) * 100)
    genes_samples[nrow(genes_samples) + 1, ] <- row
}

plot1 <- show_gene_counts(genes_samples)
png(file = paste0(FIGURES, "MM_gene_counts.png"), width = 700)
plot1
dev.off()
# nolint end
```

```{r SEQUENCE_RETRIEVAL_PREDICTED, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
# Making a data frame that stores data of how many PWM hits were found in
# Homo sapiens genome:
seq_dataset_predicted <- get_sequences(BSgenome.Hsapiens.UCSC.hg38, hg_genes)
# nolint end
```

```{r PWM_COUNTS_PREDICTED, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
pwm_hit_predicted <- count_pwm_hits(seq_dataset_predicted, pwm, "predicted")
# nolint end
```



NUO ŠIOS VIETOS YRA NEVISAI GERAI! REIKIA IMTI GENŲ SEKAS, O NE PAVADINIMUS!
PRATĘSTI TAISYTI NUO ČIA!

```{r HG_PEAK_ANNOTATION, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
hg_annotated_peaks <-
    annotate_peaks(grl_hg, grch38, TxDb.Hsapiens.UCSC.hg38.knownGene,
                   "org.Hs.eg.db")

write.table(hg_annotated_peaks,
            file = paste0(RESULTS, "/hg_annotated_peaks.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE)

hg_genes2 <- GRangesList()

for (sample in seq_along(hg_annotated_peaks)) {
    for (sample2 in 1:length(hg_genes)) {
        selected_genes <-
            tolower(unique(hg_annotated_peaks[[sample]]$gene_symbol)) %in%
            tolower(hg_genes[[sample2]]$gene_symbol)
                                
        genes <- gene_ranges[selected_genes]        
    }
    hg_genes2[[sample]] <- genes
}

```

```{r SEQUENCE_RETRIEVAL_CONTROL, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
# Making a data frame that stores data of how many PWM hits were found in
# Homo sapiens samples:
seq_dataset_control <- get_sequences(BSgenome.Hsapiens.UCSC.hg38, hg_genes2)
# nolint end
```

```{r PWM_COUNTS_CONTROL, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
pwm_hit_control <- count_pwm_hits(seq_dataset_control, pwm, "control")
# nolint end
```

PRATĘSTI NUO ČIA!!!

```{r}


a <- as.data.frame(pwm_hit_predicted[1])
b <- as.data.frame(pwm_hit_control[1])

length(rownames(a))
(length(rownames(b)) / length(rownames(a))) *100
# length(rownames(a[aa, ]))
# length(rownames(b[bb, ]))
intersecting_genes <- intersect(a[, "Gene"],b[, "Gene"])
length(intersecting_genes)


# length(rownames(a[a$"Gene" %in% intersecting_genes, ]))


aa <- a[a$"Gene" %in% unique(intersecting_genes), ]
bb <- b[b$"Gene" %in% unique(intersecting_genes), ]
head(bb)

total <- merge(aa, bb, by=c("Gene"))
total
length(rownames(aa))
length(rownames(bb))

# nolint end
```

