```{r LIBRARIES, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
library(pacman)
p_load(GenomicRanges, ggplot2, rtracklayer, dplyr, plyranges, ChIPseeker,
TxDb.Mmusculus.UCSC.mm10.knownGene, TxDb.Hsapiens.UCSC.hg38.knownGene,
BSgenome.Mmusculus.UCSC.mm10, BSgenome.Hsapiens.UCSC.hg38, org.Mm.eg.db,
org.Hs.eg.db, annotables, rBLAST, rentrez, clusterProfiler)
# nolint end
```

```{r PATH_DECLARATION, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
PROJECT     <- paste0("/home/daniele/Desktop/IV_course/II_semester/TF_analysis/")
BIGBEDS     <- paste0(PROJECT, "Input/")
MMUSCULUS   <- paste0(BIGBEDS, "Mus_musculus/")
HSAPIENS    <- paste0(BIGBEDS, "Homo_sapiens/")
INTER_FILES <- paste0(PROJECT, "Intermediate_data/")
SCRIPTS     <- paste0(PROJECT, "Scripts/")
RESULTS     <- paste0(INTER_FILES, "Generated_files")
FASTAS      <- paste0(INTER_FILES, "FASTA/")
FIGURES     <- paste0(PROJECT, "Figures/")

SAMPLES     <- read.csv(file = paste0(INTER_FILES, "sample_key.csv"))
DATABASES   <- paste0(PROJECT, "Databases/")
# nolint end
```

```{r FUNCTION_SOURCING, echo=FALSE, message=FALSE, warning=FALSE}
source(paste0(SCRIPTS, "functions.R"))
source(paste0(SCRIPTS, "plots.R"))
```

```{r BIGBED_READING, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Listing sample BigBed files for each analyzed organism:
bbfiles_mm <- list.files(path = MMUSCULUS, "*bb")
bbfiles_hg <- list.files(path = HSAPIENS, "*bb")

# Making a granges list for each analyzed organism:
grl_mm <- GRangesList()
grl_hg <- GRangesList()

chr_abr <- c(paste0("chr", 1:19), "chrX", "chrY")
len_mm <- length(bbfiles_mm)    # Length of samples for Mus musculus
len_hg <- length(bbfiles_hg)    # Length of samples for Homo sapiens

# Creating GRanges objects and extracting certain chromosomes for Mus musculus:
for (i in 1:len_mm) {
    grl_mm[[i]] <-
        import(paste0(MMUSCULUS, bbfiles_mm[i])) %>%
        dplyr::filter(seqnames %in% chr_abr)
    names(grl_mm)[i] <- SAMPLES$Graph_names[SAMPLES$Filename == bbfiles_mm[i]]
}

# Creating GRanges objects and extracting certain chromosomes for Homo sapiens.
# These files will be used as control samples to check whether TF binding sites
# were predicted successfully using Homo sapiens annotation:
for (i in 1:len_hg) {
    grl_hg[[i]] <-
        import(paste0(HSAPIENS, bbfiles_hg[i])) %>%
        dplyr::filter(seqnames %in% chr_abr)
    names(grl_hg)[i] <- SAMPLES$Graph_names[SAMPLES$Filename == bbfiles_hg[i]]
}
# nolint end
```

```{r PWM_READING, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Making a list of analyzed TF PWM matrices from different organisms:
pwm <- get_PWM("GATA3_MOUSE.H11MO.0.A.pwm")
# nolint end
```

```{r MM_PEAK_ANNOTATION, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
mm_annotated_peaks <-
    annotate_peaks(grl_mm, grcm38, TxDb.Mmusculus.UCSC.mm10.knownGene,
                   "org.Mm.eg.db")

write.table(mm_annotated_peaks,
            file = paste0(RESULTS, "/mm_annotated_peaks.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE)
# nolint end
```

```{r HS_PEAK_ANNOTATION, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
# Getting the peaks that are significant. q value must be lower than 2
# (-log10(0.01) = 2):
significant_peaks <- grl_hg[[1]][grl_hg[[1]]$q_value < 2, ] %>% list(.)

hg_annotated_peaks <-
    annotate_peaks(significant_peaks, grch38, TxDb.Hsapiens.UCSC.hg38.knownGene,
                   "org.Hs.eg.db")

write.table(hg_annotated_peaks,
            file = paste0(RESULTS, "/hg_annotated_peaks.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE)
# nolint end
```

```{r NUCLEOTIDE_TO_AMINOACIDS, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
mmgene_ranges <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)

# Adding 'gene_symbol' column to gene_ranges GRange:
mmgene_ranges$gene_symbol <-
    mapIds(org.Mm.eg.db, keys = mmgene_ranges$gene_id, column = "SYMBOL",
           keytype = "ENTREZID", multiVals = "first")

mm_genes <- GRangesList()

for (sample in seq_along(mm_annotated_peaks)) {
    selected_genes <-
        tolower(unique(mm_annotated_peaks[[sample]]$gene_symbol)) %in%
        tolower(mmgene_ranges$gene_symbol)
                            
    genes <- mmgene_ranges[selected_genes]
    mm_genes[[sample]] <- genes 
}

# Čia išgaunamos nukleotidų sekos konkretiems genams:
gene_sequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10, mm_genes[[1]])

library(ensembldb)
library(EnsDb.Mmusculus.v79)
library(deepredeff)

edb <- EnsDb.Mmusculus.v79

all_proteins <- character()
gprotein_sequences <- data.frame(matrix(ncol = 2, nrow = 0))

for (gene_name in 1:length(mm_genes[[1]]$gene_symbol)) {
    gene <- mm_genes[[1]]$gene_symbol[[gene_name]]

    if (is.na(gene)) {
        next
    } else {
        proteins <- proteins(edb, filter = GeneNameFilter(gene),
                     return.type = "AAStringSet")
    }

    if (length(proteins) == 0) {
        next
    } else {
        print(paste("Getting sequence for", gene, "..."))
        proteins_df <- aas_to_df(proteins)
        proteins_df$name <- gene

        genes_in_sample <- length(unique(mm_annotated_peaks[[sample]]$gene_symbol))
        row <- c(gene, proteins_df$seq[1]) # Galbūt čia reikėtų pataisyti, kad
                                           # imtų viską, o ne tik pirmą seką
        gprotein_sequences[nrow(gprotein_sequences) + 1, ] <- row

        all_proteins <- append(all_proteins, proteins[1])
    }
}

sink(paste0(PROJECT, "aa_sequences_80.txt"))
print(gprotein_sequences)
sink()

# nolint end
```

Su Blast surandame Homo sapiens genus, kurie yra panašūs į Mus musculus
genus (sekų indentiškumas = 60).

Sudarome lentelę, kurioje nurodytas Mus musculus genas ir jo ID bei jo
Blast atitikmuo - genas ir jo ID.




```{r BLAST_SEARCH, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Loading downloaded Homo sapiens protein database:
human_db <-
    blast(db = paste0(DATABASES,
          "Homo_sapiens_protein/GCF_000001405.38_GRCh38.p12_protein.fna"),
          type = "blastp")

# Performing blastp search using defined Homo sapiens protein database and
# 'AAStringSet' object that contains protein sequences of annotated genes:
predictions <-
    predict(human_db, all_proteins, BLAST_args = "-num_threads 10")

# Saving blastp result into a file so that it can be used for further analysis:
# FIND BETTER WAY TO SAVE THE RESULT!
sink(paste0(PROJECT, "predictions_protein.txt"))
print(predictions)
sink()

# Reading a file that stores blastp result:
predictions <- read.table(paste0(PROJECT, "predictions_p.txt"))
predictions <- predictions[predictions$`Perc.Ident` > 50, ]

grouped_predictions <-
    predictions %>%
    group_by(QueryID) %>%
    dplyr::filter(`Perc.Ident` == max(`Perc.Ident`)) %>%
    as.data.frame()

proteins_symbols <-
    ensembldb::select(EnsDb.Mmusculus.v79, keys = grouped_predictions$`QueryID`,
                      keytype = "PROTEINID", columns = c("ENTREZID", "SYMBOL"))

merged_data <-
    merge(proteins_symbols, grouped_predictions,
          by.x = "PROTEINID", by.y = "QueryID") %>%
    dplyr::select(c("PROTEINID", "ENTREZID", "SYMBOL",
                    "SubjectID", "Perc.Ident"))

# Removing '.[1-9]' from accession numbers (if the step is skipped the
# following function 'bitr()' does not return any results):
merged_data$SubjectID <-
    merged_data$SubjectID %>%
    gsub("\\.[1-9]+", "", .)

# Accessing Entrez identification numbers and gene symbols for every
# accession number (some accession numbers do not have Entrez id, therefore
# they are not matched):
entrezs_symbols <-
    bitr(merged_data$SubjectID, fromType = "ACCNUM",
         toType = c("ENTREZID", "SYMBOL"), OrgDb = org.Hs.eg.db)

# Creating a final dataframe that matches all subject and query data:
matched_data <-
    merge(merged_data, entrezs_symbols, by.x = "SubjectID", by.y = "ACCNUM") %>%
    rename("QueryID" = "PROTEINID", "EntrezQuery" = "ENTREZID.x",
           "QuerySymbol" = "SYMBOL.x", "EntrezSubject" = "ENTREZID.y",
           "SubjectSymbol" = "SYMBOL.y", "PercentIdentity" = "Perc.Ident") %>%
    dplyr::select(c("SubjectID", "QueryID", "SubjectSymbol", "QuerySymbol",
                    "EntrezSubject", "EntrezQuery", "PercentIdentity"))

# nolint end
```

```{r}
# nolint start
matched_data <- matched_data[!(is.na(matched_data$EntrezQuery)), ]
hggene_ranges <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene)

mm_sequences <- character()
hg_sequences <- character()

for (id in 1:length(matched_data$EntrezQuery)) {
    seq_m <- getSeq(BSgenome.Mmusculus.UCSC.mm10,
                    mmgene_ranges[mmgene_ranges$gene_id ==
                                  matched_data$EntrezQuery[id], ])

    if (length(seq_m) == 0) { next }
    else {
        names(seq_m) <- matched_data[id, "QuerySymbol"]
        mm_sequences <- append(mm_sequences, seq_m)
    }

    seq_h <- getSeq(BSgenome.Hsapiens.UCSC.hg38,
                    hggene_ranges[hggene_ranges$gene_id ==
                                  matched_data$EntrezSubject[id], ])

    if (length(seq_h) == 0) { next }
    else {
        names(seq_h) <- matched_data[id, "SubjectSymbol"]
        hg_sequences <- append(hg_sequences, seq_h)
    }
}



countPWM(as.matrix(pwm), as.character(hg_sequences[[8]]), min.score = "60%")
# nolint end
```


```{r}
# nolint start
percentages <- data.frame(matrix(ncol = 5, nrow = 0))
colnames(percentages) <- c("QueryGene", "SubjectGene", "QueryPWMHits",
                           "SubjectPWMHits", "Percent")


for (gene in 1:length(mm_sequences)) {
    mm_hits <-
        as.numeric(countPWM(as.matrix(pwm), as.character(mm_sequences[[gene]]),
                            min.score = "60%"))

    hg_hits <-
        as.numeric(countPWM(as.matrix(pwm), as.character(hg_sequences[[gene]]),
                            min.score = "60%"))

    row <- c(names(mm_sequences[gene]), names(hg_sequences[gene]),
             mm_hits, hg_hits, (mm_hits / hg_hits) * 100)

    percentages[nrow(percentages) + 1, ] <- row            
}

percentages$SubjectGene %in% hg_annotated_peaks[[1]]$gene_symbol



# nolint end
```














