```{r LIBRARIES, echo=FALSE, message=FALSE, warning=FALSE}
library(pacman)
p_load(GenomicRanges, ggplot2, rtracklayer, dplyr, plyranges, org.Mm.eg.db,
TxDb.Mmusculus.UCSC.mm10.knownGene, ChIPseeker, BSgenome.Mmusculus.UCSC.mm10,
annotables, BSgenome.Hsapiens.UCSC.hg19, TxDb.Hsapiens.UCSC.hg38.knownGene,
org.Hs.eg.db)
```

```{r PATH_DECLARATION, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
PROJECT     <- "/home/daniele/Desktop/IV_course/II_semester/TF_analysis/"
BIGBEDS     <- paste0(PROJECT, "Input/")
MMUSCULUS   <- paste0(BIGBEDS, "Mus_musculus/")
HSAPIENS    <- paste0(BIGBEDS, "Homo_sapiens/")
INTER_FILES <- paste0(PROJECT, "Intermediate_data/")
SCRIPTS     <- paste0(PROJECT, "Scripts/")
RESULTS     <- paste0(INTER_FILES, "Generated_files")
FIGURES     <- paste0(PROJECT, "Figures/")

SAMPLES     <- read.csv(file = paste0(INTER_FILES, "sample_key.csv"))
# nolint end
```

```{r FUNCTION_SOURCING, echo=FALSE, message=FALSE, warning=FALSE}
source(paste0(SCRIPTS, "functions.R"))
source(paste0(SCRIPTS, "plots.R"))
```

```{r BIGBED_READING, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Listing sample BigBed files for each analyzed organism:
bbfiles_mm <- list.files(path = MMUSCULUS, "*bb")
bbfiles_hg <- list.files(path = HSAPIENS, "*bb")

# Making a granges list for each analyzed organism:
grl_mm <- GRangesList()
grl_hg <- GRangesList()

chr_abr <- c(paste0("chr", 1:19), "chrX", "chrY")
len_mm <- length(bbfiles_mm)    # Length of samples for Mus musculus
len_hg <- length(bbfiles_hg)    # Length of samples for Homo sapiens

# Creating GRanges objects and extracting certain chromosomes for Mus musculus:
for (i in 1:len_mm) {
    grl_mm[[i]] <-
        import(paste0(MMUSCULUS, bbfiles_mm[i])) %>%
        filter(seqnames %in% chr_abr)
    names(grl_mm)[i] <- SAMPLES$Graph_names[SAMPLES$Filename == bbfiles_mm[i]]
}

# Creating GRanges objects and extracting certain chromosomes for Homo sapiens.
# These files will be used as control samples to check whether TF binding sites
# were predicted successfully using Homo sapiens annotation:
for (i in 1:len_hg) {
    grl_hg[[i]] <-
        import(paste0(HSAPIENS, bbfiles_hg[i])) %>%
        filter(seqnames %in% chr_abr)
    names(grl_hg)[i] <- SAMPLES$Graph_names[SAMPLES$Filename == bbfiles_hg[i]]
}
# nolint end
```

```{r PWM_READING, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Making a list of analyzed TF PWM matrices from different organisms:
pwm <- get_PWM("GATA3_MOUSE.H11MO.0.A.pwm")
# nolint end
```

```{r MM_PEAK_ANNOTATION, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start

mm_annotated_peaks <-
    annotate_peaks(grl_mm, grcm38, TxDb.Mmusculus.UCSC.mm10.knownGene,
                   "org.Mm.eg.db")

write.table(mm_annotated_peaks,
            file = paste0(RESULTS, "/mm_annotated_peaks.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE)

# nolint end
```

```{r HG_ANNOTATION, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
# Getting positions of Homo sapiens genes:
gene_ranges <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)

# Adding 'gene_symbol' column to gene_ranges GRange:
gene_ranges$gene_symbol <-
    mapIds(org.Hs.eg.db, keys = gene_ranges$gene_id, column = "SYMBOL",
           keytype = "ENTREZID", multiVals = "first")

write.table(gene_ranges,
            file = paste0(RESULTS, "/hg_annotation_genes.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE)

hg_genes <- GRangesList()

for (sample in seq_along(mm_annotated_peaks)) {
    selected_genes <-
        tolower(unique(mm_annotated_peaks[[sample]]$gene_symbol)) %in%
        tolower(gene_ranges$gene_symbol)
                            
    genes <- gene_ranges[selected_genes]

    # These are the Homo sapiens genes that overlap with annotated Mus musculus
    # peak genes. This information will be used to compare the results with
    # control Homo sapiens samples:
    hg_genes[[sample]] <- genes 
}
# nolint end
```


```{r GENE_COUNTS, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
genes_samples <- data.frame(matrix(ncol = 3, nrow = 0))
colnames(genes_samples) <- c("Sample", "Gene_count", "Percentage")

for (sample in seq_along(mm_annotated_peaks)) {
    genes_in_sample <- length(unique(mm_annotated_peaks[[sample]]$gene_symbol))
    row <-
        c(names(mm_annotated_peaks[sample]), genes_in_sample,
          (genes_in_sample / length(unique(gene_ranges$gene_symbol))) * 100)
    genes_samples[nrow(genes_samples) + 1, ] <- row
}

plot1 <- show_gene_counts(genes_samples)
png(file = paste0(FIGURES, "MM_gene_counts.png"), width = 700)
plot1
dev.off()
# nolint end
```

```{r SEQUENCE_RETRIEVAL, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
# Making a data frame that stores data of how many PWM hits were found in
# Homo sapiens genome:
seq_dataset_hg <- c()

# Creating a sequence dataset that stores Homo sapiens gene sequences:
for (object in 1:length(hg_genes)) {
    gene_sequences <- getSeq(BSgenome.Hsapiens.UCSC.hg19, hg_genes[[object]])
    names(gene_sequences) <- as.data.frame(hg_genes[[object]])$gene_symbol

    # Creating a sequence vector for all the samples:
    seq_dataset_hg <- c(seq_dataset_hg, gene_sequences)
}
# nolint end
```

```{r PWM_COUNTS, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start
pwm_hit_counts <- count_pwm_hits(seq_dataset_hg, pwm) 

# nolint end
```


PRATĘSTI NUO ŠIOS DALIES KOREGAVIMO IR FUNKCIJŲ PARAŠYMO!!!


```{r HG_PEAK_ANNOTATION, echo=FALSE, message=FALSE, warning=FALSE, output=FALSE}
# nolint start

hg_annotated_peaks <-
    annotate_peaks(grl_hg, grch38, TxDb.Hsapiens.UCSC.hg38.knownGene,
                   "org.Hs.eg.db")

write.table(hg_annotated_peaks,
            file = paste0(RESULTS, "/hg_annotated_peaks.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE)


hg_genes2 <- GRangesList()

for (sample in seq_along(hg_annotated_peaks)) {
    selected_genes <-
        tolower(unique(hg_annotated_peaks[[sample]]$gene_symbol)) %in%
        tolower(gene_ranges$gene_symbol)
                            
    genes <- gene_ranges[selected_genes]

    # These are the Homo sapiens genes that overlap with annotated Mus musculus
    # peak genes. This information will be used to compare the results with
    # control Homo sapiens samples:
    hg_genes2[[sample]] <- genes 
}


seq_dataset_hg2 <- c()

# Creating a sequence dataset that stores Homo sapiens gene sequences:
for (object in 1:length(hg_genes2)) {
    gene_sequences <- getSeq(BSgenome.Hsapiens.UCSC.hg19, hg_genes2[[object]])
    names(gene_sequences) <- as.data.frame(hg_genes2[[object]])$gene_symbol

    # Creating a sequence vector for all the samples:
    seq_dataset_hg2 <- c(seq_dataset_hg2, gene_sequences)
}
# nolint end
```

